### PR TITLE
Rework BlockLoader with new model

### PR DESCRIPTION
**Pull request by l4l**
### Description of the Change

Change BlockLoader's interface to the new shared model

### CODE REVIEW

File: shared_model/interfaces/iroha_internal/commit.hpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef IROHA_SHARED_MODEL_COMMIT_HPP
+#define IROHA_SHARED_MODEL_COMMIT_HPP
+
+#include <memory>
+#include <rxcpp/rx-observable.hpp>
```
[the remark this answers is not in the document]
Comment (l4l) [reply shared_model/interfaces/iroha_internal/commit.hpp]: Yeah, forgot to remove

File: irohad/network/impl/block_loader_impl.cpp [hunk 2]
```diff
@@ -96,7 +104,7 @@ nonstd::optional<Block> BlockLoaderImpl::retrieveBlock(
   protocol::Block block;
 
   // request block with specified hash
-  request.set_hash(block_hash.to_string());
+  request.set_hash(toBinaryString(block_hash));
```
Comment (muratovv) [inline irohad/network/impl/block_loader_impl.cpp]: What is the reason to change hashing here?

File: irohad/network/impl/block_loader_impl.cpp [hunk 3]
```diff
@@ -105,25 +113,32 @@ nonstd::optional<Block> BlockLoaderImpl::retrieveBlock(
...
-  if (not crypto_provider_->verify(result)) {
+  auto result = std::make_shared<shared_model::proto::Block>(block);
+  std::unique_ptr<iroha::model::Block> old_block(result->makeOldModel());
+  if (not crypto_provider_->verify(*old_block)) {
     log_->error(kInvalidBlockSignatures);
     return nonstd::nullopt;
   }
 
-  return result;
+  return static_cast<Commit>(result);
```
Comment (muratovv) [inline irohad/network/impl/block_loader_impl.cpp]: Also, looks like a bug.

File: irohad/network/impl/block_loader_impl.cpp [hunk 4]
```diff
@@ -105,25 +113,32 @@ nonstd::optional<Block> BlockLoaderImpl::retrieveBlock(
...
 }
 
-nonstd::optional<Peer> BlockLoaderImpl::findPeer(Peer::KeyType pubkey) {
+nonstd::optional<iroha::model::Peer> BlockLoaderImpl::findPeer(
+    PublicKey pubkey) {
```
Comment (muratovv) [inline irohad/network/impl/block_loader_impl.cpp]: Use const ref instead.

File: irohad/network/impl/block_loader_impl.cpp [hunk 5]
```diff
@@ -105,25 +113,32 @@ nonstd::optional<Block> BlockLoaderImpl::retrieveBlock(
...
     return nonstd::nullopt;
   }
 
+  auto &blob = pubkey.blob();
   auto it = std::find_if(
-      peers.value().begin(), peers.value().end(), [&pubkey](const auto &peer) {
-        return peer.pubkey == pubkey;
+      peers.value().begin(), peers.value().end(), [&blob](const auto &peer) {
+        return peer.pubkey.size() == blob.size()
+            && std::equal(peer.pubkey.begin(),
```
Comment (muratovv) [inline irohad/network/impl/block_loader_impl.cpp]: /s/&&/and

File: irohad/network/impl/block_loader_impl.cpp [hunk 6]
```diff
@@ -38,53 +45,54 @@ const char *kInvalidBlockSignatures = "Block signatures are invalid";
...
+    auto reader =
+        this->getPeerStub(peer.value()).retrieveBlocks(&context, request);
+    while (reader->Read(&block)) {
+      auto result = std::make_shared<shared_model::proto::Block>(block);
+      std::unique_ptr<iroha::model::Block> old_block(result->makeOldModel());
+      if (not crypto_provider_->verify(*old_block)) {
+        log_->error(kInvalidBlockSignatures);
+        context.TryCancel();
+      } else {
+        subscriber.on_next(static_cast<Commit>(result));
```
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/network/impl/block_loader_impl.cpp]: Due to `shared_ptr<iface>` -> `shared_ptr<proto::>`
I guess it's better use polymorphic wrapper here

File: irohad/network/impl/block_loader_impl.cpp (hunk 2 above)
Comment (l4l) [reply irohad/network/impl/block_loader_impl.cpp]: `to_string` in SM returns human-readable string, not binary

File: irohad/network/impl/block_loader_impl.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (muratovv) [reply irohad/network/impl/block_loader_impl.cpp]: What is the point to cast here? Better to return interface instead.

File: irohad/network/impl/block_loader_impl.cpp (hunk 3 above)
Comment (l4l) [reply irohad/network/impl/block_loader_impl.cpp]: That one is required, it has 2 casts. First proto->iface, then iface->optional

File: irohad/network/impl/block_loader_impl.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/network/impl/block_loader_impl.cpp]: oops, i mean reverse ofc, proto->iface

File: test/module/irohad/network/block_loader_test.cpp [hunk 7]
```diff
@@ -52,15 +60,30 @@ class BlockLoaderTest : public testing::Test {
...
+    std::copy_n(
+        peer_key.blob().begin(), peer.pubkey.size(), peer.pubkey.begin());
     peers.push_back(peer);
 
     ASSERT_TRUE(server);
     ASSERT_NE(port, 0);
   }
 
-  Peer peer;
+  auto base() const {
```
Comment (Alexey-N-Chernyshov) [inline test/module/irohad/network/block_loader_test.cpp]: mb rename to getBlockBuilder?

File: test/module/irohad/network/block_loader_test.cpp [hunk 8]
```diff
@@ -52,15 +60,30 @@ class BlockLoaderTest : public testing::Test {
...
+    constexpr auto kTotal = (1 << 5) - 1;
```
Comment (Alexey-N-Chernyshov) [inline test/module/irohad/network/block_loader_test.cpp]: Not clear. What does it mean?

File: test/module/irohad/network/block_loader_test.cpp [hunk 9]
```diff
@@ -52,15 +60,30 @@ class BlockLoaderTest : public testing::Test {
...
+    return shared_model::proto::TemplateBlockBuilder<
+               kTotal,
+               shared_model::validation::DefaultBlockValidator,
+               shared_model::proto::Block>()
+        .txNumber(0)
+        .height(0)
```
Comment (Alexey-N-Chernyshov) [inline test/module/irohad/network/block_loader_test.cpp]: you always set it to 1 further. mb it is better to set here?

File: test/module/irohad/network/block_loader_test.cpp (hunk 8 above)
Comment (l4l) [reply test/module/irohad/network/block_loader_test.cpp]: Exactly what does TOTAL mean in builder, but it's private

File: test/module/irohad/network/block_loader_test.cpp (hunk 9 above)
Comment (l4l) [reply test/module/irohad/network/block_loader_test.cpp]: Not at the all cases, look [here](https://github.com/hyperledger/iroha/pull/949/files/3405edd8544cf05eece72ebf64bbc0ffbf99eee9#diff-4ec25e0d6a030bb199ddde4f4a681be6R170) for example

File: test/module/irohad/network/block_loader_test.cpp (hunk 7 above)
Comment (l4l) [reply test/module/irohad/network/block_loader_test.cpp]: I just want to leave the name as short as possible, though maybe you're and it's not really that meaningful. I'll fix it

File: test/module/irohad/network/block_loader_test.cpp (hunk 9 above)
Comment (l4l) [reply test/module/irohad/network/block_loader_test.cpp]: upd: doesn't seem necessary, I'll fix that as well

File: libs/common/wrapper.hpp [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+   * Creates a generic wrapped type with custom-type ctor if specified
+   *
+   * @tparam Y is returning wrapped type
+   * @tparam T is ctor that should be called on Wrapper creation
+   * @tparam Args are arguments to the ctor
+   * @return created wrapped instance
+   *
+   * note: ctor without args should be called with explicit Y AND T
+   */
+  template <typename Y, typename T = Y, typename... Args>
```
Comment (muratovv) [inline libs/common/wrapper.hpp]: Wow, is it working with default and variadic types together? How compiler recognize where `makeWrapper(M, N)` deduced to `Y=M, T=N` and from other side `makeWrapper(A, B)` deduced to `Y=A, Args=[B]`

File: irohad/network/impl/block_loader_impl.cpp [hunk 11]
```diff
@@ -105,25 +115,32 @@ nonstd::optional<Block> BlockLoaderImpl::retrieveBlock(
...
-  if (not crypto_provider_->verify(result)) {
+  auto result = makeWrapper<Block, shared_model::proto::Block>(block);
+  std::unique_ptr<iroha::model::Block> old_block(result->makeOldModel());
+  if (not crypto_provider_->verify(*old_block)) {
     log_->error(kInvalidBlockSignatures);
     return nonstd::nullopt;
   }
 
-  return result;
+  return nonstd::optional<Wrapper<Block>>(std::move(result));
```
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/network/impl/block_loader_impl.cpp]: That will change iface, moreover there's [IR-960](https://soramitsu.atlassian.net/browse/IR-960)

File: libs/common/wrapper.hpp (hunk 10 above)
Comment (l4l) [reply libs/common/wrapper.hpp]: variadic template can be void

**Pull request merged**